src/awkward/_dispatch.py

```diff
@@ -0,0 +1,47 @@
+from functools import wraps
+from inspect import signature
```

Comment:
Tools from `inspect` have a habit of not working when the function is defined on the fly (i.e. not in an imported module). That would definitely be true if you were asking for the source code or line number, but I don't know about `inspect.signature`. If it's implemented using other pieces of `inspect`, it would have this problem. But I don't know if it does, since a `function.__code__` object contains all the information one would need in order to reconstruct a signature...

Reply:
> It looks like you find out if a given __awkward_function__ applies by seeing if it raises an error or lacks a __qualname__ (?); the usual way is for the function to return (not raise) NotImplemented (not NotImplementedError). Would that make sense here?

No — this is actually the client implementation (e.g. `dask-awkward`). The idea is that unlike NEP-18, we don't need to try multiple implementations to handle a function. If any array-like argument implements the Awkward protocol, it becomes responsible for handling the call (and must error if it can't). This simplifies things considerably, and I don't think that we need to support the multiple-array-type case (e.g. dask-awkward mixed with some other library). Do you feel this is too restrictive? In future, should we decide to relax this constraint, we could (and interpret `NotImplementedError` as `return NotImplemented`).

Reply:
Yes, since there's a pathway to extend this to the `return NotImplemented` pattern in the future, let's go ahead with it.

Version 1: if you write a `__awkward_function__` method, you must provide an implementation. That implementation might raise an exception, but that exception will be passed to the user and will look like an error. (I.e. we won't catch `NotImplementedError` or any other exception raised by a third-party `__awkward_function__`. We only anticipate one downstream library implementing `__awkward_function__` methods, and that's dask-awkward.

Version 2: if you write a `__awkward_function__` method, you can `return NotImplemented`, and then we'll check the other arguments for `__awkward_function__` methods. The value of this would just to be "normal," to adhere to a common pattern in the Python world.

Adding version 2 after version 1 has been out for a while wouldn't invalidate the libraries written for version 1, since they have to not `return NotImplemented` in order to adhere to the version 1 protocol. The only odd case is if `NotImplemented` was the _desired output_ of an Awkward function that a third-party library wanted to return. That's very unlikely. But we could plug that gap by adding a restriction to the version 1 protocol: "It is illegal to return `NotImplemented`." That would make it like a reserved bit.